### ISSUE TITLE
cache tier (chunkserver)

### ISSUE THREAD

**Issue by onlyjob** [lizardfs/lizardfs#270]
It could be very helpful for performance if chunkservers could place frequently accessible data to faster HDDs and move chunks between local hard disks according to cache hits. It would allow chuinkservers to perform very well on servers with many rotational HDDs and just one SSD.

**Comment by Burning-Daylight** [comment]
Intra-chunkserver caching can be _already_ achieved with bcache or ZFS' L2Arc. Performance-aware  placing and movement of chunks between chunk-servers would be absolutely fantastic though... Configuration of chunk placement would be great in general.

**Comment by onlyjob** [comment]
To some extent chunk placement can be already configured using labels and goals.
Flash caching with _bcache_, _flashcache_ and _enhanceio_ is not great and does not produce too much improvement while adding complexity. On chunkservers we have achieved more efficient acceleration with _dm-writeboost_ and ext4 external journals.

**Comment by Burning-Daylight** [comment]
> To some extent chunk placement can be already configured using labels and goals.

But if I have, let's say 3 goals: `safe: london paris _` and client in London starts writing a new file, where will the first replica go? What happens if client in London re-writes a chunk?

> On chunkservers we have achieved more efficient acceleration with dm-writeboost and ext4 external journals.

Thanks for sharing! Do I get it right that your workload or LizardFS is random-write performance sensitive? If so, did  you try ZFS and Btrfs, which usually perform good at random-writes?

**Comment by onlyjob** [comment]
> But if I have, let's say 3 goals: safe: london paris _ and client in London starts writing a new file, where will the first replica go? What happens if client in London re-writes a chunk?

I remember this was discussed once: https://github.com/lizardfs/lizardfs/issues/227#issuecomment-68253260
If I understand it properly, first/new replica will be created on random chunkserver and replicated as per its goal settings ASAP. Existing/re-written chunks will be updated simultaneously at once. Reads can be somewhat optimised by primitive _mfstopology.cfg(5)_ settings to provide hints to clients about proximity of chunkservers.

> Do I get it right that your workload or LizardFS is random-write performance sensitive?

Perhaps sometimes but not really... Distributed nature of LizardFS turns any write workload to randomised one. I'd say slow rotational HDDs do not handle any parallel access too well. Once write stress is relieved HDDs perform much better and access latency becomes noticeably lower.

Read caching of data on chunkservers is not too effective due to highly randomised distributed access patterns.

> did you try ZFS and Btrfs, which usually perform good at random-writes?

I did not try ZFS (as I do not want to introduce too much resource-intensive complexity). Also as far as I'm aware only Debian kFreeBSD have native support for ZFS and FUSE alternative is unlikely to be efficient. I'm not excited about stacking FUSE file systems.

Btrfs is good, but needs regular maintenance (re-balancing) to avoid free space issues. Btrfs feels slower than _ext4_ and _xfs_ for writing partially due to CRC and copy-on-write features. We use Btrfs with _compress=lzo_ only on some chunkservers to pin highly compressible infrequently accessible data.

**Comment by Burning-Daylight** [comment]
> first/new replica will be created on random chunkserver and replicated as per its goal settings ASAP. Existing/re-written chunks will be updated simultaneously at once.

Which, unfortunately, makes geo-replication practically unusable atm. It will happen that in the case `safe: chicago paris _`, chunkserver `paris` will be selected for the first chunk written from client in Chicago, which will reduce performance dramatically. Re-writes of the existing chunks will always be very slow. 
Having an option to respect _mfstopology.cfg_ also for writes would be great.

> Perhaps sometimes but not really...

Why do you observe efficient acceleration with _dm-writeboost_ then? If I understood correctly, _dm-writeboost_ and ext4 external journals mostly accelerate random-write?

> native support for ZFS

[Native ZFS is available](http://zfsonlinux.org/) for all major Linux distributions for a few years already, also [used](https://pthree.org/2012/04/17/install-zfs-on-debian-gnulinux/) quite successfully, from what I can tell. ZFS' embedded caching (L2Arc, ZIL) features might be beneficial.

> Btrfs feels slower

Both ZFS and Brtfs may be slower or faster then ext4/xfs depending on the workload. ZFS/Brtfs generally outperform ext4 at random writes.

**Comment by onlyjob** [comment]
> Why do you observe efficient acceleration with dm-writeboost then?

Because writing affects performance of reads on slow large capacity HDDs that do not handle parallel operations too well...

Native ZFS is still not available from official Debian repositories hence out of option.
We are pretty strict about it here. Only official packages. ;)

> ZFS/Brtfs generally outperform ext4 at random writes.

That's the first time I read about that. It would be very strange for COW file system to outperform traditional one on writes. Certainly my observation do not confirm that on Btrfs but I did not do any testing to check this particular claim. Also it might be that _ext4_ mounted with `journal_async_commit` is faster than Btrfs on writes...

**Comment by onlyjob** [comment]
> Having an option to respect `mfstopology.cfg` also for writes would be great.

Indeed. I've been already waiting for this feature for a while... :)  Perhaps we should remind more often or maybe someone could prepare a patch...

**Comment by Burning-Daylight** [comment]
> That's the first time I read about that. It would be very strange for COW file system to outperform traditional one on writes.
http://www.phoronix.com/scan.php?page=article&item=linux-44-zfs&num=2
In couple of tests ZFS is ahead, however I can not suggest this as a comprehensive study as performance is very configuration-dependent.

**Comment by FlorianHeigl** [comment]
As @onlyjob wrote, for a use case like this most standard local caching solutions do not add a large improvement. tracking the (non-checkloop!) accesses to a chunk allows to cache what is a "hot zone". 
This is what any "real" storage does, and what is currently not implemented in bcache/enhanceio/lvm/flashcache. zfs is a very large hammer for this problem.
Also please keep in mind it less efficient to that end than having some cache layer at the distributed fs level:
You have lower hitrate and slower cache warmup - even if one chunkserver already cached this block it will still be uncached on another.

The other bad mistake is to read through the cache at all times (i.e. flashcache, the first version of the ceph cache tier) or blindly skip sequentatial IO.
